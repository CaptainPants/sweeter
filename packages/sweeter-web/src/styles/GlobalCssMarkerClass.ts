import type { StylesheetDependencyProvider } from './StylesheetDependencyProvider.js';
import type {
    AbstractGlobalCssClass,
    AbstractGlobalCssStylesheet,
    GlobalStyleSheetContentGeneratorContext,
} from './types.js';

export interface GlobalCssMarkerClassOptions {
    /**
     * Base the class name for this class on this value (it will be prefixed/suffixed or otherwise made unique).
     */
    className: string;
    dependencies?: StylesheetDependencyProvider;
}

/**
 * A classname that doesn't have a definition, for use in the definition of other classes. This could be e.g. 'primary' or 'outline'.
 */
export class GlobalCssMarkerClass implements AbstractGlobalCssClass {
    public readonly className: string;
    public readonly id: string;
    public readonly symbol: symbol;

    #dependencies?: AbstractGlobalCssStylesheet[];

    constructor(options: GlobalCssMarkerClassOptions);
    constructor({ className, dependencies }: GlobalCssMarkerClassOptions) {
        this.className = className;
        this.id = className;
        this.symbol = Symbol('GlobalCssClass-' + className);

        if (dependencies) {
            dependencies.addDependencyListener((dependency) => {
                this.#dependencies ??= [];
                this.#dependencies.push(dependency);
            });
        }
    }

    getContent(
        context: GlobalStyleSheetContentGeneratorContext,
    ): string | undefined {
        return undefined;
    }

    getReferencedStylesheets():
        | readonly AbstractGlobalCssStylesheet[]
        | undefined {
        return this.#dependencies;
    }
}
