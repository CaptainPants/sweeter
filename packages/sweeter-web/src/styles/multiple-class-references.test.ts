import { stylesheet, GlobalCssStylesheet } from './index.js';
import { GlobalCssMarkerClass } from './GlobalCssMarkerClass.js';

it('Multiple class references produce sane output', () => {
    const classA = new GlobalCssMarkerClass({ className: 'ClassA' });
    const classB = new GlobalCssMarkerClass({ className: 'ClassB' });

    const sheet = new GlobalCssStylesheet({
        content: stylesheet`
            .${classA} {
                color: red;
            }
            .${classB} {
                color: green;

                &:hover {
                    color: blue;
                }
            }`,
        id: 'test',
    });

    const result = sheet.getContent({
        getPrefixedClassName: (classObject) => classObject.className,
    });

    expect(result).toMatchSnapshot();
});
