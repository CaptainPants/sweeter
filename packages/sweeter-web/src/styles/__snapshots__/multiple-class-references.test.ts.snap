// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Multiple class references produce sane output 1`] = `
".ClassA {
    color: red;
}
.ClassB {
    color: green;
}
.ClassB:hover {
    color: blue;
}
"
`;
